File: code/archive/flex-cell.lua
function __getObjectPosition__(a,b)
    -- compatibility routine, wrong results could be returned in some situations, in CoppeliaSim <4.0.1
    if b==sim.handle_parent then
        b=sim.getObjectParent(a)
    end
    if (b~=-1) and (sim.getObjectType(b)==sim.object_joint_type) and (sim.getInt32Param(sim.intparam_program_version)>=40001) then
        a=a+sim.handleflag_reljointbaseframe
    end
    return sim.getObjectPosition(a,b)
end

function __compute_xyz_flexcell__(X,Y,Z)
        Z_table_level = 0.020 
        YMIN = 0
        YMAX = 23
        XMIN = 0
        XMAX = 15
        HOLE_DIST = 0.05
        x_min = -0.072
        x_max = 1.128
        y_max = 0.731 - 0.020 -- compensation
        y_min = -0.369 - 0.020 -- compensation
        comp_x = x_max - ((Y+1)*HOLE_DIST) 
        comp_y = (y_max - ((X)*HOLE_DIST))
        comp_z = Z_table_level + Z * HOLE_DIST
        return comp_x,comp_y,comp_z
end

function __compute_inverse_xyz_flexcell__(x,y,z)
        Z_table_level = 0.020
        YMIN = 0
        YMAX = 23
        XMIN = 0
        XMAX = 15
        HOLE_DIST = 0.05
        x_min = -0.072
        x_max = 1.128
        y_max = 0.731 - 0.020 -- compensation
        y_min = -0.369 - 0.020 -- compensation

        Y = (x_max - x)/HOLE_DIST - 1
        X = (y_max - y)/HOLE_DIST
        Z = (z - Z_table_level)/HOLE_DIST
        return math.floor(X+0.5),math.floor(Y+0.5),math.floor(Z+0.5)
end

function __compute_yz_joint__(yc,zc)
        theta = math.atan(zc/yc)
        alpha = math.pi/4 + theta
        zj = (math.pow(math.pow(zc,2) + math.pow(yc,2),(1/2))) * math.cos(math.pi/2 - alpha)
        yj = (math.pow(math.pow(zc,2) + math.pow(yc,2),(1/2))) * math.sin(math.pi/2 - alpha)
        return yj,zj
end

function __compute_inverse_yz_joint__(yj,zj)
        if ((yj == 0.0) or (zj == 0.0)) then
            return yj,zj
        end
        theta = math.atan(yj/zj)
        alpha = math.pi/4 + theta
        zc = (math.pow(math.pow(zj,2) + math.pow(yj,2),(1/2))) * math.sin(math.pi/2 - alpha)
        yc = (math.pow(math.pow(zj,2) + math.pow(yj,2),(1/2))) * math.cos(math.pi/2 - alpha)
        return yc,zc
end

function sysCall_init()
    filename = 'test.csv'
    basepath = '/home/santiago/York/flex-cell/'
    sim.pauseSimulation()
    base_name = "/1_base_link_collision/"
    j0 = sim.getObject(base_name .. "11_shoulder_pan_joint") -- J0
    j1 = sim.getObject(base_name .. "18_shoulder_lift_joint") -- J1
    j2 = sim.getObject(base_name .."14_elbow_joint") -- J2
    j3 = sim.getObject(base_name .. "12_wrist_1_joint") -- J3
    j4 = sim.getObject(base_name .. "19_wrist_2_joint") -- J4
    j5 = sim.getObject(base_name .. "13_wrist_3_joint") -- J5
    
    l_fg = sim.getObject(base_name .. "16_leftFingerGripper") -- Left finger gripper (OnRobot 2FG7)
    r_fg = sim.getObject(base_name .. "15_rightFingerGripper") -- Right finger gripper (OnRobot 2FG7)
    
    print("Sim objects correctly loaded")
    

    simRemoteApi.start(8888)
    print("Start client as 'simple_bullet_remoteApi 8888 "..j0.." "..j1.." "..j2.." "..j3.." "..j4.." "..j5.." "..l_fg.." "..r_fg.."'")
    h = 0
    lifting = false
    end_effector = sim.getObject(base_name .. "8_link_2fg7_base_8__link_2fg7_box")
    positions = {}
    path = basepath..'tests/'..filename
    idx = 1

    for line in io.lines(path) do

        X,Y,Z = line:match("([^,]+),([^,]+),([^,]+)")
        n_X = tonumber(X)
        n_Y = tonumber(Y)
        n_Z = tonumber(Z)
        print(X..","..Y..","..Z)
        
        -- Perform the transformations with existing models
        --q0 =
        --q1 = 
        --q2 = 
        --q3 = 
        --q4 = 
        --q5 =
        x,yc,zc = __compute_xyz_flexcell__(n_X,n_Y,n_Z)
        yj,zj = __compute_yz_joint__(yc,zc)
	--ee_replica = sim.copyPasteObjects({end_effector},0)[1]


        sim.setObjectPosition(end_effector,-1, {x, yj, zj})
        sim.setObjectOrientation(end_effector,-1, {-5*math.pi/4, 0, 0})
        idx = idx + 1
        --positions[idx] = ee_replica
        positions[idx] = {x, yj, zj,-5*math.pi/4, 0, 0}
    end
    
end

function sysCall_actuation()
    -- put your actuation code here
    
end

function sysCall_sensing()
    -- put your sensing code here
end

function sysCall_cleanup()
    -- do some clean-up here
    output_file = basepath..'results/'..filename
    output = io.open(output_file, "w")
    for i,o in pairs(positions) do
        --pos = sim.getObjectPosition(o,-1)
        --ori = sim.getObjectOrientation(o,-1)
        --print(pos[1], pos[2], pos[3], ori[1],ori[2],ori[3])
        --output:write(pos[1],",",pos[2],",",pos[3],",",ori[1],",",ori[2],",",ori[3],"\n")
        print(o[1], o[2], o[3], o[4],o[5],o[6])
        output:write(o[1],",",o[2],",",o[3],",",o[4],",",o[5],",",o[6],"\n")
    end
    output:close()


    simRemoteApi.stop(8888)
    print("Stopped remote api")
    
end

-- See the user manual or the available code snippets for additional callback functions and details

File: code/flex-cell.py
#python
from robots_flexcell import robots
import csv
import numpy as np


basepath = '/home/santiago/York/flex-cell/'
filename = 'test.csv'
positions = {}
discrete_positions = {}
step_size = 0.5
current_time = 0.0


ur_robot_model = robots.UR5e_RoboSim_Simulation(mqtt_enabled=False,zmq_enabled=False)
ur_robot_model.set_motion_time(2.0)## Definition of the trajectory time


def sysCall_init():
    # do some initialization here
    #sim = require('sim')
    global j0
    global j1
    global j2
    global j3
    global j4
    global j5
    global l_fg
    global r_fg
    global end_effector
    global ur5e_sim
    global ref_cuboid

    simBase = sim.getObject('.')
    #simTip = sim.getObject('./tip')
    #simTarget = sim.getObject('./target')

    #sim.pauseSimulation()
    base_name = "/1_base_link_collision/"
    base_name_alt = "/1_base_link_collision/"
    j0 = sim.getObject(base_name + "17_shoulder_pan_joint") # J0
    j1 = sim.getObject(base_name + "11_shoulder_lift_joint") # J1
    j2 = sim.getObject(base_name +"16_elbow_joint") # J2
    j3 = sim.getObject(base_name + "15_wrist_1_joint") # J3
    j4 = sim.getObject(base_name + "18_wrist_2_joint") # J4
    j5 = sim.getObject(base_name + "14_wrist_3_joint") # J5

    #j0_alt = sim.getObject(base_name_alt + "17_shoulder_pan_joint") # J0
    #j1_alt = sim.getObject(base_name_alt + "15_shoulder_lift_joint") # J1
    #j2_alt = sim.getObject(base_name_alt +"14_elbow_joint") # J2
    #j3_alt = sim.getObject(base_name_alt + "12_wrist_1_joint") # J3
    #j4_alt = sim.getObject(base_name_alt + "11_wrist_2_joint") # J4
    #j5_alt = sim.getObject(base_name_alt + "18_wrist_3_joint") # J5
    r_fg = sim.getObject(base_name + "13_rightFingerGripper") # Right finger gripper (OnRobot 2FG7)
    l_fg = sim.getObject(base_name + "12_leftFingerGripper") # Left finger gripper (OnRobot 2FG7)


    #l_fg_alt = sim.getObject(base_name_alt + "11_leftFingerGripper") # Left finger gripper (OnRobot 2FG7)
    #r_fg_alt = sim.getObject(base_name_alt + "15_rightFingerGripper") # Right finger gripper (OnRobot 2FG7)



    ur5e_sim = sim.getObject("/1_base_link_collision")
    ref_cuboid = sim.getObject("/Cuboid")
    print("Sim objects correctly loaded")








    simRemoteApi.start(8888)
    print("Start client as 'simple_bullet_remoteApi 8888 "+ str(j0)+" "+str(j1)+" "+str(j2)+" "+str(j3)+" "+str(j4)+" "+str(j5)+" "+str(l_fg)+" "+str(r_fg)+"'")


    n_iterations = 0
    MAX_ITERATIONS = 1
    idx = 0
    path = basepath+'tests/'+filename
    with open(path,'r') as file:
        reader = csv.reader(file, delimiter=",")
        for row in reader:
            if n_iterations == MAX_ITERATIONS:
                continue
            n_iterations = n_iterations + 1
            X = row[0]
            Y = row[1]
            Z = row[2]
            n_X = int(X)
            n_Y = int(Y)
            n_Z = int(Z)
            discrete_positions[idx] = [n_X,n_Y,n_Z]
            print(str(X)+","+str(Y)+","+str(Z))
            # Transformations
            #n_X -> n_Z
            #n_Y -> -n_Y
            #n_Z -> n_X


            x,yc,zc = ur_robot_model.compute_xyz_flexcell(n_X,n_Y,Z=n_Z)
            yj,zj = ur_robot_model.compute_yz_joint(yc,zc)
            # Perform the transformations with existing models
            target_position = ur_robot_model.compute_q(x,yj,zj)[0]
            q_traj = ur_robot_model.compute_trajectory(ur_robot_model.get_previous_position(),target_position,plot=False)
            q_traj = q_traj
            ur_robot_model.transmit_robot_motion(q_traj)

            n_entry = round((step_size+current_time)/0.05)
            motion_done = False
            if n_entry>len(q_traj):
                n_entry = len(q_traj)
                motion_done = True
            n_entry = len(q_traj)
            actual_traj_array = q_traj[n_entry-1]
            actual_q0 = actual_traj_array[0]
            actual_q1 = actual_traj_array[1]
            actual_q2 = actual_traj_array[2]
            actual_q3 = actual_traj_array[3]
            actual_q4 = actual_traj_array[4]
            actual_q5 = actual_traj_array[5]
            print("moving to [" + str(actual_traj_array) + "]")


            sim.setJointTargetPosition(j0,actual_q0)
            sim.setJointTargetPosition(j1,actual_q1)
            sim.setJointTargetPosition(j2,actual_q2)
            sim.setJointTargetPosition(j3,actual_q3)
            sim.setJointTargetPosition(j4,actual_q4)
            sim.setJointTargetPosition(j5,actual_q5)

            #sim.setJointTargetPosition(j1_alt,np.pi/2)
            #sim.setJointTargetPosition(j2_alt,np.pi/2)
            #sim.setJointTargetPosition(j0_alt,actual_q0)
            #sim.setJointTargetPosition(j1_alt,actual_q1)
            #sim.setJointTargetPosition(j2_alt,actual_q2)
            #sim.setJointTargetPosition(j3_alt,actual_q3)
            #sim.setJointTargetPosition(j4_alt,actual_q4)
            #sim.setJointTargetPosition(j5_alt,actual_q5)
            #sim.setObjectPose(end_effector,ref_cuboid,[x,yj,zc,-5*np.pi/4,0,0,0])

            #time.sleep(step_size)
            #current_time = current_time + step_size
            if motion_done:
                previous_array = np.array([actual_q0,actual_q1,actual_q2,actual_q3,actual_q4,actual_q5])
                ur_robot_model.set_previous_position(previous_array)
                #continue
            idx = idx + 1
            positions[idx] = [x, yj, zj]



def sysCall_actuation():
    # put your actuation code here
    '''
    global j0
    global j1
    global j2
    global j3
    global j4
    global j5
    global l_fg
    global r_fg
    global end_effector
    global ur5e_sim
    global ref_cuboid

    #for i,pos in discrete_positions.items():
    #n_X = pos[0]
    #n_Y = pos[1]
    #n_Z = pos[2]
    n_X = discrete_positions[1][0]
    n_Y = discrete_positions[1][1]
    n_Z = discrete_positions[1][2]
    x,yc,zc = ur_robot_model.compute_xyz_flexcell(n_X,n_Y,Z=n_Z)
    yj,zj = ur_robot_model.compute_yz_joint(yc,zc)
    # Perform the transformations with existing models
    target_position = ur_robot_model.compute_q(x,yj,zj)[0]
    q_traj = ur_robot_model.compute_trajectory(ur_robot_model.get_previous_position(),target_position,plot=False)
    q_traj = q_traj
    ur_robot_model.transmit_robot_motion(q_traj)

    n_entry = round((step_size+current_time)/0.05)
    motion_done = False
    if n_entry>len(q_traj):
        n_entry = len(q_traj)
        motion_done = True
    actual_traj_array = q_traj[n_entry-1]
    actual_q0 = actual_traj_array[0]
    actual_q1 = actual_traj_array[1]
    actual_q2 = actual_traj_array[2]
    actual_q3 = actual_traj_array[3]
    actual_q4 = actual_traj_array[4]
    actual_q5 = actual_traj_array[5]

    sim.setJointTargetPosition(j0,actual_q0)
    sim.setJointTargetPosition(j1,actual_q1)
    sim.setJointTargetPosition(j2,actual_q2)
    sim.setJointTargetPosition(j3,actual_q3)
    sim.setJointTargetPosition(j4,actual_q4)
    sim.setJointTargetPosition(j5,actual_q5)
    print("Moving joints to: [" + str(actual_q0) + "," + str(actual_q1) + "," +str(actual_q2) + "," + str(actual_q3) + "," + str(actual_q4) + "," + str(actual_q5) + "]")
    #sim.setObjectPose(end_effector,ref_cuboid,[x,yj,zc,-5*np.pi/4,0,0,0])

    #time.sleep(step_size)
    #current_time = current_time + step_size
    if motion_done:
        previous_array = np.array([actual_q0,actual_q1,actual_q2,actual_q3,actual_q4,actual_q5])
        ur_robot_model.set_previous_position(previous_array)
    positions[1] = [x, yj, zj]
    #time.sleep(5.0)'''
    pass

def sysCall_sensing():
    # put your sensing code here
    pass

def sysCall_cleanup():
    # do some clean-up here
    global j0
    global j1
    global j2
    global j3
    global j4
    global j5
    global l_fg
    global r_fg
    global end_effector
    global ur5e_sim
    global ref_cuboid


    output_file = basepath+'results/'+filename
    with open(output_file, "w") as output:
        for i,o in positions.items():
            print(str(o[0]) + ", " + str(o[1]) + ","  + str(o[2]))
            output.write(str(o[0])+","+str(o[1])+","+str(o[2])+"\n")
        output.close()


    simRemoteApi.stop(8888)
    print("Stopped remote api")

# See the user manual or the available code snippets for additional callback functions and details
